File: apps/patter-voice-sdk/libraries/typescript/src/services/ivr.ts
/**
 * IVR auto-navigation activity for telephony calls (TypeScript port).
 *
 * Detects IVR prompts via transcribed speech, forwards DTMF responses
 * through `CallControl.sendDtmf`, and recovers from two common failure
 * modes:
 *
 * 1. The agent hears the same IVR prompt repeated several times
 *    (loop detection). `TfidfLoopDetector` flags this by comparing the
 *    cosine similarity of recent transcript chunks.
 * 2. The IVR falls silent while both parties are passive (silence
 *    detection). A debounced timer triggers a follow-up after
 *    `maxSilenceDuration` seconds of combined silence.
 *
 * The Python port uses scikit-learn for TF-IDF; TypeScript has no
 * equivalent battle-tested package in the std library, so we ship a
 * minimal in-house bag-of-words + cosine-similarity implementation.
 * It is intentionally simple — enough to match repeated IVR prompts.
 */

import type { CallControl } from "../metrics";
import { getLogger } from "../logger";

// ---------------------------------------------------------------------------
// DTMF event taxonomy
// ---------------------------------------------------------------------------

/** Valid DTMF tone values (keypad characters). */
export const DTMF_EVENTS = [
  "1", "2", "3", "4", "5", "6", "7", "8", "9", "0",
  "*", "#", "A", "B", "C", "D",
] as const;

/** Single DTMF tone value (a member of `DTMF_EVENTS`). */
export type DtmfEvent = (typeof DTMF_EVENTS)[number];

/** Join DTMF events into a space-separated debug string. */
export function formatDtmf(events: DtmfEvent[]): string {
  return events.join(" ");
}

// ---------------------------------------------------------------------------
// Bag-of-words + cosine similarity (minimal TF-IDF-ish loop detector)
// ---------------------------------------------------------------------------

const WORD_RE = /[a-z0-9]+/g;

function tokenize(text: string): string[] {
  return (text.toLowerCase().match(WORD_RE) ?? []);
}

/** Term-frequency bag-of-words vector. */
function bagOfWords(text: string): Map<string, number> {
  const counts = new Map<string, number>();
  for (const token of tokenize(text)) {
    counts.set(token, (counts.get(token) ?? 0) + 1);
  }
  return counts;
}

function cosineSimilarity(a: Map<string, number>, b: Map<string, number>): number {
  if (a.size === 0 || b.size === 0) return 0;

  let dot = 0;
  // Iterate the smaller map for efficiency.
  const [small, big] = a.size <= b.size ? [a, b] : [b, a];
  for (const [term, weight] of small) {
    const other = big.get(term);
    if (other !== undefined) {
      dot += weight * other;
    }
  }

  let normA = 0;
  for (const w of a.values()) normA += w * w;
  let normB = 0;
  for (const w of b.values()) normB += w * w;

  const denom = Math.sqrt(normA) * Math.sqrt(normB);
  if (denom === 0) return 0;
  return dot / denom;
}

// ---------------------------------------------------------------------------
// TfidfLoopDetector
// ---------------------------------------------------------------------------

/** Constructor options for `TfidfLoopDetector`. */
export interface TfidfLoopDetectorOptions {
  /** Number of recent chunks to keep in the comparison window. */
  windowSize?: number;
  /** Cosine similarity above which two chunks are "the same prompt". */
  similarityThreshold?: number;
  /** Consecutive near-duplicates required before firing. */
  consecutiveThreshold?: number;
}

/**
 * Detects repeated IVR prompts via cosine similarity on bag-of-words
 * vectors. Not a full TF-IDF implementation — good enough for catching
 * IVRs that re-read the same menu.
 */
export class TfidfLoopDetector {
  private readonly windowSize: number;
  private readonly similarityThreshold: number;
  private readonly consecutiveThreshold: number;
  private chunks: Array<{ text: string; vec: Map<string, number> }> = [];
  private consecutiveSimilar = 0;

  constructor(opts: TfidfLoopDetectorOptions = {}) {
    const {
      windowSize = 20,
      similarityThreshold = 0.85,
      consecutiveThreshold = 3,
    } = opts;

    if (windowSize <= 0) {
      throw new Error("windowSize must be greater than 0");
    }
    if (similarityThreshold < 0 || similarityThreshold > 1) {
      throw new Error("similarityThreshold must be between 0.0 and 1.0");
    }
    if (consecutiveThreshold <= 0) {
      throw new Error("consecutiveThreshold must be greater than 0");
    }

    this.windowSize = windowSize;
    this.similarityThreshold = similarityThreshold;
    this.consecutiveThreshold = consecutiveThreshold;
  }

  /** Forget all previously observed chunks and reset the consecutive-hit counter. */
  reset(): void {
    this.chunks = [];
    this.consecutiveSimilar = 0;
  }

  /** Record a new transcript chunk in the rolling window. */
  addChunk(text: string): void {
    this.chunks.push({ text, vec: bagOfWords(text) });
    if (this.chunks.length > this.windowSize) {
      this.chunks = this.chunks.slice(-this.windowSize);
    }
  }

  /** Returns true once the most recent chunks look like a repeated IVR prompt. */
  checkLoopDetection(): boolean {
    if (this.chunks.length < 2) return false;

    const last = this.chunks[this.chunks.length - 1];
    let maxSim = 0;
    for (let i = 0; i < this.chunks.length - 1; i++) {
      const sim = cosineSimilarity(last.vec, this.chunks[i].vec);
      if (sim > maxSim) maxSim = sim;
    }

    if (maxSim > this.similarityThreshold) {
      this.consecutiveSimilar += 1;
    } else {
      this.consecutiveSimilar = 0;
    }

    return this.consecutiveSimilar >= this.consecutiveThreshold;
  }
}

// ---------------------------------------------------------------------------
// Debounced silence helper
// ---------------------------------------------------------------------------

class DebouncedCall {
  private timer: ReturnType<typeof setTimeout> | null = null;
  constructor(
    private readonly callback: () => Promise<void> | void,
    private readonly delayMs: number,
  ) {}

  schedule(): void {
    this.cancel();
    this.timer = setTimeout(() => {
      this.timer = null;
      Promise.resolve(this.callback()).catch((err) => {
        getLogger().error("IVR silence callback raised:", err);
      });
    }, this.delayMs);
  }

  cancel(): void {
    if (this.timer !== null) {
      clearTimeout(this.timer);
      this.timer = null;
    }
  }
}

// ---------------------------------------------------------------------------
// IVRActivity
// ---------------------------------------------------------------------------

/** Async callback fired when the TF-IDF detector trips. */
export type LoopCallback = () => Promise<void> | void;
/** Async callback fired after sustained silence. */
export type SilenceCallback = () => Promise<void> | void;

/** Constructor options for `IVRActivity`. */
export interface IVRActivityOptions {
  /** Seconds of combined silence before firing `onSilence`. Default `5.0`. */
  maxSilenceDuration?: number;
  /** Enable the TF-IDF loop detector. Default `true`. */
  loopDetector?: boolean;
  /** Fired when the loop detector trips. */
  onLoopDetected?: LoopCallback;
  /** Fired after `maxSilenceDuration` seconds of combined silence. */
  onSilence?: SilenceCallback;
}

/** OpenAI-style function tool spec with attached handler. */
export interface IVRToolDefinition {
  name: string;
  description: string;
  parameters: {
    type: "object";
    properties: Record<string, unknown>;
    required?: string[];
  };
  handler: (args: { events: string[] }) => Promise<string>;
}

/**
 * Coordinate IVR navigation heuristics for a single call.
 *
 * Usage::
 *
 *     const ivr = new IVRActivity(callControl);
 *     await ivr.start();
 *
 *     // In the STT loop, on each final transcript:
 *     await ivr.onUserTranscribed(text);
 *
 *     // When done:
 *     await ivr.stop();
 */
export class IVRActivity {
  private readonly callControl: CallControl;
  private readonly maxSilenceDurationMs: number;
  private readonly loopDetector: TfidfLoopDetector | null;
  private readonly onLoopDetected?: LoopCallback;
  private readonly onSilence?: SilenceCallback;

  private currentUserState: string | null = null;
  private currentAgentState: string | null = null;
  private readonly debouncedSilence: DebouncedCall;
  private lastShouldSchedule: boolean | null = null;
  private started = false;

  constructor(callControl: CallControl, opts: IVRActivityOptions = {}) {
    const {
      maxSilenceDuration = 5.0,
      loopDetector = true,
      onLoopDetected,
      onSilence,
    } = opts;

    this.callControl = callControl;
    this.maxSilenceDurationMs = maxSilenceDuration * 1000;
    this.loopDetector = loopDetector ? new TfidfLoopDetector() : null;
    this.onLoopDetected = onLoopDetected;
    this.onSilence = onSilence;

    this.debouncedSilence = new DebouncedCall(
      () => this.onSilenceDetected(),
      this.maxSilenceDurationMs,
    );
  }

  /** Begin tracking transcripts and silence; call once per call. */
  async start(): Promise<void> {
    this.started = true;
  }

  /** Stop tracking and cancel any pending silence timer. */
  async stop(): Promise<void> {
    this.debouncedSilence.cancel();
    this.started = false;
  }

  /** Feed a final user-side transcript chunk into the loop detector. */
  async onUserTranscribed(text: string): Promise<void> {
    if (!this.started || !text) return;

    if (this.loopDetector !== null) {
      this.loopDetector.addChunk(text);
      if (this.loopDetector.checkLoopDetection()) {
        this.loopDetector.reset();
        if (this.onLoopDetected) {
          try {
            await this.onLoopDetected();
          } catch (err) {
            getLogger().error("IVR onLoopDetected callback raised:", err);
          }
        }
      }
    }
  }

  /** Record the current user-turn state (e.g. `"listening"`, `"away"`). */
  noteUserState(state: string): void {
    // Guard on the lifecycle flag: a state transition reported after
    // ``stop()`` re-armed the silence timer and fired ``onSilence`` on a
    // dead call.
    if (!this.started) return;
    this.currentUserState = state;
    this.scheduleSilenceCheck();
  }

  /** Record the current agent-turn state (e.g. `"idle"`, `"listening"`). */
  noteAgentState(state: string): void {
    if (!this.started) return;
    this.currentAgentState = state;
    this.scheduleSilenceCheck();
  }

  /** Tool definitions to expose to the LLM (currently only `send_dtmf_events`). */
  get tools(): IVRToolDefinition[] {
    return [this.buildSendDtmfTool()];
  }

  // -------- internals --------

  private scheduleSilenceCheck(): void {
    const shouldSchedule = this.shouldScheduleCheck();
    if (shouldSchedule) {
      if (this.lastShouldSchedule) return;
      this.debouncedSilence.schedule();
    } else {
      this.debouncedSilence.cancel();
    }
    this.lastShouldSchedule = shouldSchedule;
  }

  private shouldScheduleCheck(): boolean {
    const userSilent =
      this.currentUserState === "listening" || this.currentUserState === "away";
    const agentSilent =
      this.currentAgentState === "idle" || this.currentAgentState === "listening";
    return userSilent && agentSilent;
  }

  private async onSilenceDetected(): Promise<void> {
    if (this.onSilence) {
      try {
        await this.onSilence();
      } catch (err) {
        getLogger().error("IVR onSilence callback raised:", err);
      }
    }
  }

  private buildSendDtmfTool(): IVRToolDefinition {
    const allowedValues = [...DTMF_EVENTS];

    const handler = async (args: { events: string[] }): Promise<string> => {
      const events = args.events ?? [];
      const validated: string[] = [];
      for (const raw of events) {
        if (!allowedValues.includes(raw as DtmfEvent)) {
          return `Failed to send DTMF event: invalid digit '${raw}'`;
        }
        validated.push(raw);
      }

      const digits = validated.join("");
      try {
        if (!this.callControl.sendDtmf) {
          return "Failed to send DTMF events: CallControl.sendDtmf not implemented";
        }
        await this.callControl.sendDtmf(digits, { delayMs: 300 });
      } catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        return `Failed to send DTMF events: ${msg}`;
      }
      return `Successfully sent DTMF events: ${validated.join(" ")}`;
    };

    return {
      name: "send_dtmf_events",
      description:
        "Send a list of DTMF events to the telephony provider. " +
        "Call when the IVR is asking for keypad input.",
      parameters: {
        type: "object",
        properties: {
          events: {
            type: "array",
            description: "Ordered list of DTMF digits to send.",
            items: {
              type: "string",
              enum: allowedValues,
            },
          },
        },
        required: ["events"],
      },
      handler,
    };
  }
}

File: apps/patter-voice-sdk/libraries/python/getpatter/services/ivr.py
"""
IVR auto-navigation activity for telephony calls.

Detects IVR prompts via transcribed speech, forwards DTMF responses
through ``CallControl.send_dtmf``, and recovers from two common
failure modes:

1. The agent hears the same IVR prompt repeated several times (loop
   detection). ``TfidfLoopDetector`` flags this by comparing the
   cosine similarity of recent transcript chunks.
2. The IVR falls silent while both parties are passive (silence
   detection). A debounced timer triggers a follow-up action after
   ``max_silence_duration`` seconds of combined silence.

Transcripts are pushed in explicitly by the caller (typically
``PipelineStreamHandler``) through ``on_user_transcribed``.

Optional extras install::

    pip install "getpatter[ivr]"
"""

from __future__ import annotations

import asyncio
import logging
from enum import Enum
from typing import TYPE_CHECKING, Any, Awaitable, Callable

if TYPE_CHECKING:
    from getpatter.models import CallControl

logger = logging.getLogger("getpatter.ivr")


# ── DTMF event taxonomy ─────────────────────────────────────────────────


class DtmfEvent(str, Enum):
    """Valid DTMF tones. ``value`` is the literal keypad character.

    Kept as a string enum so the value doubles as the digit string
    forwarded to ``CallControl.send_dtmf``.
    """

    ONE = "1"
    TWO = "2"
    THREE = "3"
    FOUR = "4"
    FIVE = "5"
    SIX = "6"
    SEVEN = "7"
    EIGHT = "8"
    NINE = "9"
    ZERO = "0"
    STAR = "*"
    POUND = "#"
    A = "A"
    B = "B"
    C = "C"
    D = "D"


DTMF_EVENTS: tuple[str, ...] = tuple(e.value for e in DtmfEvent)
"""All valid DTMF digit strings, matching the TypeScript ``DTMF_EVENTS`` array."""


def format_dtmf(events: list[DtmfEvent]) -> str:
    """Join DTMF events into a space-separated debug string."""
    return " ".join(event.value for event in events)


# ── Debounced silence detector ──────────────────────────────────────────


class _DebouncedCall:
    """Schedule an async callback after a quiet period.

    Reset each time ``schedule()`` is called; the most recent schedule
    "wins". ``cancel()`` stops any pending call.
    """

    def __init__(
        self,
        callback: Callable[[], Awaitable[None]],
        delay: float,
    ) -> None:
        self._callback = callback
        self._delay = delay
        self._task: asyncio.Task[None] | None = None

    def schedule(self) -> None:
        self.cancel()
        self._task = asyncio.create_task(self._run())

    def cancel(self) -> None:
        if self._task is not None and not self._task.done():
            self._task.cancel()
        self._task = None

    async def _run(self) -> None:
        try:
            await asyncio.sleep(self._delay)
        except asyncio.CancelledError:
            return

        try:
            await self._callback()
        except asyncio.CancelledError:
            raise
        except Exception:  # noqa: BLE001
            logger.exception("IVR silence callback raised")


# ── TF-IDF loop detector ────────────────────────────────────────────────


class TfidfLoopDetector:
    """Detect repeated IVR prompts via TF-IDF cosine similarity.

    Args:
        window_size: Number of recent chunks to retain for comparison.
        similarity_threshold: Cosine similarity above which two chunks
            are considered "the same prompt". Range ``[0.0, 1.0]``.
        consecutive_threshold: How many consecutive near-duplicates
            must appear before ``check_loop_detection`` returns ``True``.

    Requires the optional ``ivr`` extras (``scikit-learn`` + ``numpy``).
    """

    def __init__(
        self,
        window_size: int = 20,
        similarity_threshold: float = 0.85,
        consecutive_threshold: int = 3,
    ) -> None:
        if window_size <= 0:
            raise ValueError("window_size must be greater than 0")
        if similarity_threshold < 0.0 or similarity_threshold > 1.0:
            raise ValueError("similarity_threshold must be between 0.0 and 1.0")
        if consecutive_threshold <= 0:
            raise ValueError("consecutive_threshold must be greater than 0")

        # Lazy import — keep sklearn/numpy out of the hot path import.
        try:
            import numpy as np  # noqa: F401
            from sklearn.feature_extraction.text import TfidfVectorizer  # noqa: F401
            from sklearn.metrics.pairwise import cosine_similarity  # noqa: F401
        except ImportError as exc:
            raise ImportError(
                "TfidfLoopDetector requires the 'ivr' extras. "
                "Install with: pip install 'getpatter[ivr]'"
            ) from exc

        self._window_size = window_size
        self._similarity_threshold = similarity_threshold
        self._consecutive_threshold = consecutive_threshold
        self._chunks: list[str] = []
        self._consecutive_similar = 0

    def reset(self) -> None:
        """Clear buffered history and consecutive counter."""
        self._chunks = []
        self._consecutive_similar = 0

    def add_chunk(self, chunk: str) -> None:
        """Append a transcript chunk, trimming to the window size."""
        self._chunks.append(chunk)
        if len(self._chunks) > self._window_size:
            self._chunks = self._chunks[-self._window_size :]

    def check_loop_detection(self) -> bool:
        """Return ``True`` when the last chunk has repeated enough times."""
        if len(self._chunks) < 2:
            return False

        import numpy as np
        from sklearn.feature_extraction.text import TfidfVectorizer
        from sklearn.metrics.pairwise import cosine_similarity

        vectorizer = TfidfVectorizer()
        try:
            doc_matrix = vectorizer.fit_transform(self._chunks)
        except ValueError:
            # Empty vocabulary — all chunks were stop-words/empty.
            return False

        doc_similarity = cosine_similarity(doc_matrix)
        # Compare the newest chunk to its IMMEDIATE predecessor only:
        # ``max(last_row)`` over the whole window counted similarity to ANY
        # earlier chunk as "consecutive", so alternating prompts (A/B/A/B —
        # a normal two-question IVR) false-fired the loop detector.
        prev_similarity = doc_similarity[-1][-2] if doc_matrix.shape[0] > 1 else 0.0

        if prev_similarity > self._similarity_threshold:
            self._consecutive_similar += 1
        else:
            self._consecutive_similar = 0

        return self._consecutive_similar >= self._consecutive_threshold


# ── IVRActivity ─────────────────────────────────────────────────────────

# Callback signature: async () -> None
SilenceCallback = Callable[[], Awaitable[None]]
LoopCallback = Callable[[], Awaitable[None]]


class IVRActivity:
    """Coordinate IVR navigation heuristics for a single call.

    The activity does not own a timer loop while idle; silence checks
    are debounced and re-armed each time ``note_user_silent`` /
    ``note_agent_idle`` is called. Call sites (typically
    ``PipelineStreamHandler``) push transcripts and state updates into
    the activity; the activity reacts asynchronously.

    Args:
        call_control: The active call's control interface — used to
            forward DTMF digits.
        max_silence_duration: Seconds of combined quiet before the
            silence callback fires. Defaults to ``5.0``.
        loop_detector: Whether to enable TF-IDF loop detection.
            Requires the ``ivr`` extras. Defaults to ``True``.
        on_loop_detected: Async callback fired when the loop detector
            trips (e.g. re-prompt the IVR).
        on_silence: Async callback fired after ``max_silence_duration``
            of combined silence (e.g. speak a filler).

    Usage::

        call_control = ...  # provided by the handler
        ivr = IVRActivity(call_control)
        await ivr.start()

        # In the handler's STT loop:
        async def on_transcript(text: str) -> None:
            await ivr.on_user_transcribed(text)

        # When done:
        await ivr.stop()
    """

    def __init__(
        self,
        call_control: CallControl,
        *,
        max_silence_duration: float = 5.0,
        loop_detector: bool = True,
        on_loop_detected: LoopCallback | None = None,
        on_silence: SilenceCallback | None = None,
    ) -> None:
        self._call_control = call_control
        self._max_silence_duration = max_silence_duration
        self._loop_detector: TfidfLoopDetector | None = None
        if loop_detector:
            self._loop_detector = TfidfLoopDetector()
        self._on_loop_detected = on_loop_detected
        self._on_silence = on_silence

        self._current_user_state: str | None = None
        self._current_agent_state: str | None = None
        self._debounced_silence = _DebouncedCall(
            self._on_silence_detected,
            max_silence_duration,
        )
        self._last_should_schedule: bool | None = None
        self._started = False

    # ── lifecycle ────────────────────────────────────────────────────

    async def start(self) -> None:
        """Mark the activity as active. Idempotent."""
        self._started = True

    async def stop(self) -> None:
        """Cancel any pending silence timer. Idempotent."""
        self._debounced_silence.cancel()
        self._started = False

    # Backwards-compat alias for callers that prefer ``aclose``.
    aclose = stop

    # ── external event hooks ─────────────────────────────────────────

    async def on_user_transcribed(self, text: str) -> None:
        """Feed a final user transcript to the activity.

        Called by ``PipelineStreamHandler`` whenever the STT emits a
        final transcript. A non-final transcript must NOT be passed.
        """
        if not self._started:
            return
        if not text:
            return

        if self._loop_detector is not None:
            self._loop_detector.add_chunk(text)
            if self._loop_detector.check_loop_detection():
                logger.debug("IVRActivity: speech loop detected")
                self._loop_detector.reset()
                if self._on_loop_detected is not None:
                    try:
                        await self._on_loop_detected()
                    except Exception:  # noqa: BLE001
                        logger.exception("IVR on_loop_detected callback raised")

    def note_user_state(self, state: str) -> None:
        """Record the user's latest voice activity state.

        ``state`` is one of ``"listening"``, ``"speaking"``, ``"away"``;
        any other value is treated as "active".
        """
        self._current_user_state = state
        self._schedule_silence_check()

    def note_agent_state(self, state: str) -> None:
        """Record the agent's latest state (``"idle"``, ``"speaking"``, ...)."""
        self._current_agent_state = state
        self._schedule_silence_check()

    # ── internals ────────────────────────────────────────────────────

    def _schedule_silence_check(self) -> None:
        should_schedule = self._should_schedule_check()
        if should_schedule:
            # Only (re)schedule if we weren't already waiting.
            if self._last_should_schedule:
                return
            self._debounced_silence.schedule()
        else:
            self._debounced_silence.cancel()
        self._last_should_schedule = should_schedule

    def _should_schedule_check(self) -> bool:
        is_user_silent = self._current_user_state in {"listening", "away"}
        is_agent_silent = self._current_agent_state in {"idle", "listening"}
        return is_user_silent and is_agent_silent

    async def _on_silence_detected(self) -> None:
        logger.debug("IVRActivity: silence detected")
        if self._on_silence is not None:
            try:
                await self._on_silence()
            except Exception:  # noqa: BLE001
                logger.exception("IVR on_silence callback raised")

    # ── LLM tool surface ─────────────────────────────────────────────

    @property
    def tools(self) -> list[dict[str, Any]]:
        """Return LLM function-call tool specs bound to this call.

        The returned dicts follow the OpenAI function-calling schema
        and include a ``handler`` coroutine that drives
        ``CallControl.send_dtmf``.
        """
        return [self._build_send_dtmf_tool()]

    def _build_send_dtmf_tool(self) -> dict[str, Any]:
        allowed_values = [e.value for e in DtmfEvent]

        async def handler(events: list[str]) -> str:
            # Validate each event — reject anything outside the enum.
            validated: list[str] = []
            for raw in events:
                if raw not in allowed_values:
                    return f"Failed to send DTMF event: invalid digit '{raw}'"
                validated.append(raw)

            digits = "".join(validated)
            try:
                await self._call_control.send_dtmf(digits, delay_ms=300)
            except Exception as exc:  # noqa: BLE001
                return f"Failed to send DTMF events: {exc}"
            return f"Successfully sent DTMF events: {' '.join(validated)}"

        return {
            "name": "send_dtmf_events",
            "description": (
                "Send a list of DTMF events to the telephony provider. "
                "Call when the IVR is asking for keypad input."
            ),
            "parameters": {
                "type": "object",
                "properties": {
                    "events": {
                        "type": "array",
                        "description": "Ordered list of DTMF digits to send.",
                        "items": {
                            "type": "string",
                            "enum": allowed_values,
                        },
                    }
                },
                "required": ["events"],
            },
            "handler": handler,
        }


__all__ = [
    "DTMF_EVENTS",
    "DtmfEvent",
    "IVRActivity",
    "TfidfLoopDetector",
    "format_dtmf",
]
